<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#A-little-peek-at-Lemmas" data-toc-modified-id="A-little-peek-at-Lemmas-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>A little peek at Lemmas</a></span><ul class="toc-item"><li><span><a href="#You-can-get-meaning-information-directly" data-toc-modified-id="You-can-get-meaning-information-directly-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>You can get meaning information directly</a></span></li><li><span><a href="#Multiple-lemma-names" data-toc-modified-id="Multiple-lemma-names-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Multiple lemma names</a></span></li><li><span><a href="#Grammatical-roles" data-toc-modified-id="Grammatical-roles-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Grammatical roles</a></span></li></ul></li><li><span><a href="#Use-cases" data-toc-modified-id="Use-cases-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Use cases</a></span><ul class="toc-item"><li><span><a href="#Find-words-containing..." data-toc-modified-id="Find-words-containing...-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Find words containing...</a></span></li><li><span><a href="#Substring" data-toc-modified-id="Substring-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Substring</a></span></li><li><span><a href="#Pattern-(regular-expression)" data-toc-modified-id="Pattern-(regular-expression)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Pattern (regular expression)</a></span></li><li><span><a href="#palindromes" data-toc-modified-id="palindromes-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>palindromes</a></span><ul class="toc-item"><li><span><a href="#Which-of-these-are-(or-rather-&quot;can-be&quot;)-a-verb?" data-toc-modified-id="Which-of-these-are-(or-rather-&quot;can-be&quot;)-a-verb?-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Which of these are (or rather "can be") a verb?</a></span></li><li><span><a href="#Making-functions-operable" data-toc-modified-id="Making-functions-operable-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>Making functions operable</a></span></li></ul></li><li><span><a href="#Only-p,-q,-b,-d,-and-vowels" data-toc-modified-id="Only-p,-q,-b,-d,-and-vowels-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Only p, q, b, d, and vowels</a></span><ul class="toc-item"><li><span><a href="#Containing-i,-e,-p-in-that-order,-with-other-letters-in-between" data-toc-modified-id="Containing-i,-e,-p-in-that-order,-with-other-letters-in-between-2.5.1"><span class="toc-item-num">2.5.1&nbsp;&nbsp;</span>Containing i, e, p in that order, with other letters in between</a></span></li></ul></li><li><span><a href="#S-words" data-toc-modified-id="S-words-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>S-words</a></span></li></ul></li><li><span><a href="#Search-expansion" data-toc-modified-id="Search-expansion-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Search expansion</a></span></li><li><span><a href="#Acquisition-of-word-phonetics-from-kaggle" data-toc-modified-id="Acquisition-of-word-phonetics-from-kaggle-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Acquisition of word phonetics from kaggle</a></span></li><li><span><a href="#Random-Poking-around" data-toc-modified-id="Random-Poking-around-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Random Poking around</a></span></li></ul></div>

# A little peek at Lemmas

In [110]:
from lexis import Lemmas

lm = Lemmas()
len(lm)

147306

`lm` is a `Mapping` (think "acts like a (read-only) dict")

In [112]:
from typing import Mapping

isinstance(lm, Mapping)

True

Let's have a look at a few keys

In [139]:
list(lm)[44630:44635]

['blond', 'kaunda', 'peacetime', 'intolerantly', "'hood"]

And the value of a `lm` item?

In [140]:
lm['blond']

{'blond.n.01': WordnetElement('blond.n.01'),
 'blond.n.02': WordnetElement('blond.n.02'),
 'blond.a.01': WordnetElement('blond.a.01')}

Okay, it looks like it's different meanings of "blond". The middle letter tells us its grammatical role it's a noun (`n`) or an adjective (`a`). More on that later. 

And what's a `WordnetElement`?

Well, it's another Mapping, apparently:

In [146]:
isinstance(lm['blond']['blond.n.01'], Mapping)

True

In [147]:
list(lm['blond']['blond.n.01'])

['also_sees',
 'instance_hypernyms',
 'verb_groups',
 'entailments',
 'region_domains',
 'substance_holonyms',
 'part_holonyms',
 'examples',
 'part_meronyms',
 'hyponyms',
 'member_meronyms',
 'offset',
 'causes',
 'definition',
 'lemma_names',
 'lexname',
 'member_holonyms',
 'in_topic_domains',
 'lemmas',
 'topic_domains',
 'max_depth',
 'hypernym_distances',
 'name',
 'attributes',
 'hypernyms',
 'min_depth',
 'usage_domains',
 'in_region_domains',
 'instance_hyponyms',
 'in_usage_domains',
 'similar_tos',
 'root_hypernyms',
 'pos',
 'frame_ids',
 'hypernym_paths',
 'substance_meronyms']

Wow! That's a lot of information! 

Let's look at what the definition of `'blond.n.01'` is:

In [152]:
print(lm['blond']['blond.n.01']['definition'])

a person with fair skin and hair


... actually, let's just poke at all of them (at least those that are non-empty)

In [151]:
meaning = 'blond.n.01'
print(f"Values for meaning: {meaning}")
for k, v in lm['blond'][meaning].items():
    if v:
        print(f"- {k}: {v}")

Values for meaning: blond.n.01
- hyponyms: [WordnetElement('peroxide_blond.n.01'), WordnetElement('platinum_blond.n.01'), WordnetElement('towhead.n.01')]
- offset: 9860506
- definition: a person with fair skin and hair
- lemma_names: ['blond', 'blonde']
- lexname: noun.person
- lemmas: [KvLemma('blond.n.01.blond'), KvLemma('blond.n.01.blonde')]
- max_depth: 7
- hypernym_distances: {(WordnetElement('physical_entity.n.01'), 6), (WordnetElement('entity.n.01'), 7), (WordnetElement('physical_entity.n.01'), 3), (WordnetElement('entity.n.01'), 4), (WordnetElement('living_thing.n.01'), 3), (WordnetElement('object.n.01'), 5), (WordnetElement('blond.n.01'), 0), (WordnetElement('organism.n.01'), 2), (WordnetElement('causal_agent.n.01'), 2), (WordnetElement('whole.n.02'), 4), (WordnetElement('person.n.01'), 1)}
- name: blond.n.01
- hypernyms: [WordnetElement('person.n.01')]
- min_depth: 4
- root_hypernyms: [Synset('entity.n.01')]
- pos: n
- hypernym_paths: [[WordnetElement('entity.n.01'), WordnetE

## You can get meaning information directly

What if you made a list of these strings like `'blond.n.01'`, `'blond.a.01'`... and you wanted to access the `WordnetElement` instances with all that cool information about those specifics meanings?

You could do `lm['blond']['blond.n.01']`, `lm['blond']['blond.a.01']`... But then you'd have to remember the full references `('blond', 'blond.n.01')`, `('blond', 'blond.a.01')`... 

You don't need to go through `lm['blond']` to get to the `WordnetElement` instance that gives you access to the meaning information -- you can use the `Synsets` store (i.e. Mapping). 

Note: "synset" is what Wordnet calls this. We'll just call is meaning for simplicity. I hope the purists won't mind.


In [160]:
from lexis import Synsets

In [161]:
meanings = Synsets()
meanings['blond.n.01']

WordnetElement('blond.n.01')

We saw earlier that we had `147306` lemmas (i.e. "words" or more precisely "terms"... but really precisely, "lemmas"). 

Well, we have `117659` synsets (i.e. "meanings") in the `Synsets` instance.

In [162]:
len(meanings)

117659

## Multiple lemma names

`'lemma_names'` are different ways that the same meaning can be written. 

In [153]:
lm['blond']['blond.n.01']['lemma_names']

['blond', 'blonde']

Indeed, `lm['blond']` and `lm['blonde']` really point to the same thing.

In [136]:
lm['blond']

{'blond.n.01': WordnetElement('blond.n.01'),
 'blond.n.02': WordnetElement('blond.n.02'),
 'blond.a.01': WordnetElement('blond.a.01')}

In [137]:
lm['blonde']

{'blond.n.01': WordnetElement('blond.n.01'),
 'blond.n.02': WordnetElement('blond.n.02'),
 'blond.a.01': WordnetElement('blond.a.01')}

## Grammatical roles

What are the different grammatical roles that are used in the meaning identifiers (aka synset keys) of our lemmas?

In [145]:
from collections import Counter
import re
from lexis import Lemmas

lm = Lemmas()

p_middle_of_dot_path = re.compile('(?P<first>[^\.]+)\.(?P<middle>\w+)\.(?P<last>[^\.]+)')

def extract_grammatical_role_from_meaning(meaning):
    m = p_middle_of_dot_path.match(meaning)
    if m:
        return m.groupdict().get('middle', None) 
    else:
        return None
    

c = Counter()
for meanings in lm.values():
    for meaning in meanings:
        c.update(extract_grammatical_role_from_meaning(meaning))
        
c.most_common()

[('n', 148478),
 ('v', 42751),
 ('s', 20895),
 ('a', 9846),
 ('r', 5619),
 ('_', 29),
 ('e', 28),
 ('u', 17),
 ('g', 17),
 ('i', 15),
 ('t', 14),
 ('p', 8),
 ('b', 7),
 ('o', 7),
 ('l', 6),
 ('d', 4),
 ('c', 2),
 ('m', 1),
 ('k', 1)]

# Use cases

## Find words containing...

In [163]:
from dol import filt_iter, cached_keys, add_ipython_key_completions, kvhead
from lexis import Lemmas

In [4]:
lm = Lemmas()

def print_definitions(words):
    for word in words:
        print(f"- {word}")
        for k, v in lm[word].items():
            print(f"    {'.'.join(k.split('.')[1:])}: {v['definition']}")
        
len(lm)

147306

## Substring

In [6]:
from lexis import print_word_definitions

In [5]:
substr = 'iep'
words = list(filter(lambda w: substr in w, lm))
len(words)

12

In [6]:
print_definitions(words)

- hemiepiphyte
    n.01: a plant that is an epiphyte for part of its life
- antiepileptic
    n.01: a drug used to treat or prevent convulsions (as in epilepsy)
- pieplant
    n.01: long pinkish sour leafstalks usually eaten cooked and sweetened
- liepaja
    n.01: a city of southwestern Latvia on the Baltic Sea
- semiepiphyte
    n.01: a plant that is an epiphyte for part of its life
- archiepiscopal
    a.01: of or associated with an archbishop
- tiepin
    n.01: a pin used to hold the tie in place
- giovanni_battista_tiepolo
    n.01: Italian painter (1696-1770)
- tiepolo
    n.01: Italian painter (1696-1770)
- antiepileptic_drug
    n.01: a drug used to treat or prevent convulsions (as in epilepsy)
- dnieper
    n.01: a river that rises in Russia near Smolensk and flowing south through Belarus and Ukraine to empty into the Black Sea
- dnieper_river
    n.01: a river that rises in Russia near Smolensk and flowing south through Belarus and Ukraine to empty into the Black Sea


## Pattern (regular expression)

In [2]:
import re
from lexis import Lemmas

lm = Lemmas()

def print_definitions(words):
    for word in words:
        print(f"- {word}")
        for k, v in lm[word].items():
            print(f"    {'.'.join(k.split('.')[1:])}: {v['deffffinition']}")
        

## palindromes

In [42]:
import re
from lexis import Lemmas

lm = Lemmas()

is_palindrome_with_at_least_3_letters = lambda w: len(w) >= 3 and w == w[::-1]
print(*filter(is_palindrome_with_at_least_3_letters, lm), sep=', ')

ono, waw, tot, kkk, ldl, anna, tenet, mom, igigi, sus, hallah, sls, pcp, mam, ofo, ene, alula, oto, civic, cfc, 101, tet, kazak, sss, ctc, aba, tevet, ara, wnw, mum, siris, tebet, tut-tut, ccc, naan, xix, tnt, peep, tut, kook, xanax, ala, eve, level, xxx, dud, aaa, dad, tdt, odo, pip, tibit, iii, sas, wow, radar, madam, yay, dmd, poop, ana, sos, bib, pop, isi, eye, gag, gig, cdc, dod, nun, pep, mym, bob, malayalam, sis, www, utu, non, ewe, aga, akka, noon, ese, rotor, ded, ppp, kayak, pap, wsw, pup, minim, nan, tat, ada, boob, mem, deed, nauruan, ma'am, succus, seles, cbc, tit, dvd, refer, toot


In [22]:
# Wait a minute... Where's racecar?!?
assert 'racecar' not in lm
assert 'race_car' in lm

### Which of these are (or rather "can be") a verb?

What are the keys of the lemmas? 

Answer: Synset keys -- that is, an id that references a unit of meaning

In [21]:
# what do are the values of the lemmas?
list(lm['eat'])

['eat.v.01',
 'eat.v.02',
 'feed.v.06',
 'eat.v.04',
 'consume.v.05',
 'corrode.v.01']

That little `v` seems to be indicating that the meaning is... verbal?

Let's make a function to grab that middle part of the dot path and use it to make a `is_a_verb` (more like "can be a verb"). 

In [100]:
from collections import Counter
import re
from lexis import Lemmas

lm = Lemmas()

p_middle_of_dot_path = re.compile('(?P<first>[^\.]+)\.(?P<middle>\w+)\.(?P<last>[^\.]+)')

def _extract_middle(string):
    m = p_middle_of_dot_path.match(string)
    if m:
        return m.groupdict().get('middle', None) 
    else:
        return None
    
def grammatical_roles(lemma):
    return Counter(map(_extract_middle, lm[lemma]))
    

assert grammatical_roles('go') == Counter({'n': 4, 'v': 30, 'a': 1})  # the lemma "go" can be a verb, noun, or adjective

def is_a_verb(lemma):
    return 'v' in grammatical_roles(lemma)
    
assert is_a_verb('go')
assert not is_a_verb('chess')  # unlike go, chess cannot be used as a verb, apparently

palindromes that are verbs

In [102]:
list(filter(lambda x: is_a_verb(x) and is_palindrome_with_at_least_3_letters(x), lm))

['tot',
 'tut-tut',
 'peep',
 'tut',
 'level',
 'pip',
 'wow',
 'bib',
 'pop',
 'eye',
 'gag',
 'bob',
 'kayak',
 'pup',
 'tat',
 'boob',
 'refer',
 'toot']

### Making functions operable

Making functions have arithmetic to make building a filter function more easy (and more readable)

In [108]:
from dataclasses import dataclass
from functools import wraps
from typing import Callable
import operator as o


@dataclass
class FuncsOp:
    op: Callable
    func1: Callable
    func2: Callable
        
    def __call__(self, *args, **kwargs):
        return self.op(self.func1(*args, **kwargs), self.func2(*args, **kwargs))



@dataclass
class OperableFunc:
    func: Callable = None
        
    def __post_init__(self):
        wraps(self.func)(self)
        
    def __call__(self, *args, **kwargs):
        return self.func(*args, **kwargs)
        
    def __and__(self, other):
        if isinstance(other, OperableFunc):
            other = other.func
        if self.func is None:  # then this is a neuteral element, yielding all ops to operand
            return OperableFunc(other)
        return OperableFunc(FuncsOp(o.__and__, self.func, other))
    
    def __or__(self, other):
        if isinstance(other, OperableFunc):
            other = other.func
        if self.func is None:
            return OperableFunc(other)
        return OperableFunc(FuncsOp(o.__or__, self.func, other))
    
    def __eq__(self, other):
        if isinstance(other, OperableFunc):
            other = other.func
        if self.func is None:
            return OperableFunc(other)
        return OperableFunc(FuncsOp(o.__eq__, self.func, other))
    
# ATTEMPT TO AUTOMATE THIS (but doesn't work -- has to do with "context locals" 
# (See https://werkzeug.palletsprojects.com/en/1.0.x/local/))
#     @staticmethod
#     def _mk_op_func_method(op):
#         def _op_func(self, other):
#             if isinstance(other, OperableFunc):
#                 other = other.func
#             if self.func is None:
#                 return OperableFunc(other)
#             return OperableFunc(FuncsOp(op, self.func, other))

#     for _name in ['__or__', '__eq__']:
#         locals()[_name] = _mk_op_func_method(getattr(o, _name))
        
#     del _name
    
more_than_3_letters = OperableFunc(lambda x: len(x) >= 3)
is_palindrome = OperableFunc(lambda w: w == w[::-1])

filt = (more_than_3_letters & is_palindrome)
assert filt('bob')
assert not filt('bobo')
assert not filt('mm')

# note how `is_a_verb` is not an OperableFunc, but since the left hand side is, it still works!
list(filter(more_than_3_letters & is_palindrome & is_a_verb, lm))

['tot',
 'tut-tut',
 'peep',
 'tut',
 'level',
 'pip',
 'wow',
 'bib',
 'pop',
 'eye',
 'gag',
 'bob',
 'kayak',
 'pup',
 'tat',
 'boob',
 'refer',
 'toot']

`OperableFunc()` is a "neutral element", but it will make it so that the left hand side is always an `OperableFunc` that can therefore cast the right hand side to an `OperableFunc`. 

Understood? No? Well just look at the example below where `f`, `g`, and `is_a_verb` are normal functions, and yet, since we put an `OperableFunc()` in the beginning, it works! 

In [109]:
f = lambda x: len(x) >= 3
g = lambda w: w == w[::-1]
list(filter(OperableFunc() & f & g  & is_a_verb, lm))

['tot',
 'tut-tut',
 'peep',
 'tut',
 'level',
 'pip',
 'wow',
 'bib',
 'pop',
 'eye',
 'gag',
 'bob',
 'kayak',
 'pup',
 'tat',
 'boob',
 'refer',
 'toot']

## Only p, q, b, d, and vowels

In [7]:
import re
from lexis import Lemmas

lm = Lemmas()

consonants = 'pqbd'
vowels = 'aeiou'  # 'aeiouy'
filt = re.compile(f'[{vowels}{consonants}]+$').match  # the pattern

words = list(filter(lambda w: 2 <= len(w)  <= 7, # number of letters constraing
                    filter(filt, lm)))  # filter for iep pattern
len(words)

249

In [29]:
print(*words, sep='\t')

bod	aaa	add	poa	pop	beaded	aqua	pib	edda	doob	boa	doi	padded	iodide	bop	edo	bide	eb	bai	quid	de	ade	daba	pid	baba	paba	bi	abb	bebop	pa	poop	pb	dea	odo	pope	dad	pup	bode	quad	bb	be	ea	epee	bid	pu	pique	iii	pod	bee	pub	ddi	id	baobab	equid	padua	pipidae	opaque	pappa	uppp	uub	qepiq	bibbed	adp	ada	pied	aoudad	qed	pupa	bedaub	bd	dba	bopeep	oboe	ado	eq	bpi	aid	bud	dodo	abo	qaeda	aa	pad	papua	baa	abode	bad	adad	adapid	papaia	db	bede	ai	po	doei	pep	eib	dubai	epi	boob	uuq	io	beep	quip	ad	babu	ded	ia	dud	da	qi	paid	peep	doodad	beda	eddo	boo	padda	ipidae	deep	dope	ied	doped	dopa	ii	iaea	uda	dd	baud	dido	ebb	epa	bodied	pap	ed	peba	bed	audio	deed	idea	apoidea	beau	up	pda	iud	ip	diode	bida	pi	apidae	bead	odd	od	dia	baddie	iaa	ape	ipo	dod	idp	ee	ie	daub	duo	boidae	poe	abed	adobe	pea	dude	do	aided	obi	ido	pipe	pe	doe	aiai	pd	baboo	quipu	pood	papio	equidae	iop	qadi	ab	dado	dub	adobo	bap	pei	baeda	equip	dupe	aqaba	bob	ba	dead	dada	adapa	pee	opepe	pob	iou	duad	doodia	dab	aide	pip	dipped	bubo	pipa	poi	apia

In [31]:
print_definitions(['adapa'])

- adapa
    n.01: a Babylonian demigod or first man (sometimes identified with Adam)


### Containing i, e, p in that order, with other letters in between

In [7]:
filt = re.compile('\w{0,2}i\w{0,2}e\w{0,2}p\w{0,2}$').match  # The *i*e*p* pattern

words = list(filter(lambda w: len(w) <= 6, # no more than 6 letters
                    filter(filt, lm)))  # filter for iep pattern
print_definitions(words)

9

In [12]:
print(*words, sep=', ')

tie_up, ginep, lineup, inept, pileup, tiepin, biceps, icecap, ice_up


In [9]:
print_definitions(words)

- tie_up
    v.01: secure with or as if with ropes
    v.02: invest so as to make unavailable for other purposes
    v.03: restrain from moving or operating normally
    v.01: secure in or as if in a berth or dock
    v.05: finish the last row
- ginep
    n.01: tropical American tree bearing a small edible fruit with green leathery skin and sweet juicy translucent pulp
- lineup
    n.01: (baseball) a list of batters in the order in which they will bat
    n.02: a line of persons arranged by police for inspection or identification
- inept
    s.04: not elegant or graceful in expression
    s.02: generally incompetent and ineffectual
    s.03: revealing lack of perceptiveness or judgment or finesse
- pileup
    n.01: multiple collisions of vehicles
- tiepin
    n.01: a pin used to hold the tie in place
- biceps
    n.01: any skeletal muscle having two origins (but especially the muscle that flexes the forearm)
- icecap
    n.01: a mass of ice and snow that permanently covers a large area

## S-words

Words that start with `s` but if you remove `s`, it's still a word.


In [13]:
from lexis import Lemmas
lm = Lemmas()
swords = list(filter(lambda x: x.startswith('s') and x[1:] in lm, lm))  # one line!

In [20]:
print(len(t))
print(*t[:40], sep=', ')

711
softener, spock, scent, spark, sbe, stickweed, screaky, salt, salp, sec, strap, sliver, slack, swish, sebs, sarawak, scuttle, stripping, swell, stole, spine, space, scar, sass, sewer, spitting, serving, sew, stalk, smite, sniffy, stripe, slake, stone, slit, sea, shoe, sweeper, swear_off, swan


In [35]:
from dol import filt_iter, wrap_kvs, KvReader
from lexis import Lemmas
lm = Lemmas()

@filt_iter(filt=lambda x: x.startswith('s') and x[1:] in lm)
class Swords(Lemmas):
    def __getitem__(self, k):
        v = super().__getitem__(k)
        for kk, vv in v.items():
            yield f"    {'.'.join(kk.split('.')[1:])}: {vv['definition']}"
            
s = Swords()
len(s)

711

In [36]:
k, v = s.head()
list(v)

['    n.01: a substance added to another to make it less hard']

In [38]:
from itertools import islice
for k, v in islice(s.items(), 5):
    print(f"------------ {k} -------------")
    print(*v, sep='\n')


------------ softener -------------
    n.01: a substance added to another to make it less hard
------------ spock -------------
    n.01: United States pediatrician whose many books on child care influenced the upbringing of children around the world (1903-1998)
------------ scent -------------
    n.02: a distinctive odor that is pleasant
    n.02: an odor left in passing by which a person or animal can be traced
    n.01: any property detected by the olfactory system
    v.01: cause to smell or be smelly
    v.02: catch the scent of; get wind of
    v.02: apply perfume to
------------ spark -------------
    n.01: a momentary flash of light
    n.01: merriment expressed by a brightness or gleam or animation of countenance
    n.05: electrical conduction through a gas in an applied electric field
    n.04: a small but noticeable trace of some quality that might become stronger
    n.05: Scottish writer of satirical novels (born in 1918)
    n.06: a small fragment of a burning substan

# Search expansion

Keyword: Query expansion.

In the case of single (or few words) search corpus, it's not the query being expanded, but the searched items. 

We'd like to, for example, be able to find airplane noises (which may appear as airplane, airplain, plane, airplane flying, avion...), but also more specific (propeller airplane?) or more general (vehicle?).

Here, the terms are tags that were used to annotate small segments of audio in a huge audio dataset (the audio is private and not available here). 

There's 5,179 unique terms, annotating 5,733,090 audio segments. 

The (term) data is dirty: There's spelling mistakes and inconsistencies, different languages, multiple (comma separated) tags, sometimes underscores and sometimes spaces are used to separate words, and more than 36% are empty or with a general (and automatic) `_unusual` tag. 


In [1]:
from odat.mdat.iatis import Dacc  # pip install odat

dacc = Dacc()

In [2]:
dacc.tag_counts

                         1797059
_unusual                  313948
background                118438
normal                    116777
good_engine_1_and_2       106673
                          ...   
tram_screech, traffic          1
voix                           1
water_intake_start             1
voices, bell                   1
ghhf                           1
Length: 5179, dtype: int64

In [3]:
dacc.tag_counts.sum()

5733090

# Acquisition of word phonetics from kaggle

In [1]:
from haggle import KaggleDatasets  # pip install haggle

rootdir = '/D/Dropbox/_odata/kaggle/zips'

s = KaggleDatasets(rootdir)

if 'rtatman/english-word-frequency' in s:
    del s['rtatman/english-word-frequency']  # just to prepare for the demo

list(s)

['uciml/human-activity-recognition-with-smartphones',
 'sitsawek/phonetics-articles-on-plos']

In [2]:
results = s.search('word frequency')
print(f"{len(results)=}")
list(results)[:10]

len(results)=180


['rtatman/english-word-frequency',
 'yekenot/fasttext-crawl-300d-2m',
 'rtatman/japanese-lemma-frequency',
 'rtatman/glove-global-vectors-for-word-representation',
 'averkij/lingtrain-hungarian-word-frequency',
 'lukevanhaezebrouck/subtlex-word-frequency',
 'facebook/fatsttext-common-crawl',
 'facebook/fasttext-wikinews',
 'facebook/fasttext-english-word-vectors-including-subwords',
 'kushtej/kannada-word-frequency']

In [3]:
v = s['rtatman/english-word-frequency']
type(v)

py2store.slib.s_zipfile.ZipReader

In [4]:
list(v)

['unigram_freq.csv']

In [5]:
list(s)

['uciml/human-activity-recognition-with-smartphones',
 'rtatman/english-word-frequency',
 'sitsawek/phonetics-articles-on-plos']

# Random Poking around

The following is just scrap -- has no order or declared sense -- still it's kept around be cause the code could inspire something more structured...

In [32]:
from nltk.corpus import wordnet as wn
ddir = lambda o: [x for x in dir(o) if not x.startswith('_')]
print(*ddir(wn), sep='\t')

ADJ	ADJ_SAT	ADV	MORPHOLOGICAL_SUBSTITUTIONS	NOUN	VERB	abspath	abspaths	all_lemma_names	all_synsets	citation	custom_lemmas	encoding	ensure_loaded	fileids	get_version	ic	jcn_similarity	langs	lch_similarity	lemma	lemma_count	lemma_from_key	lemmas	license	lin_similarity	morphy	of2ss	open	path_similarity	readme	res_similarity	root	ss2of	synset	synset_from_pos_and_offset	synset_from_sense_key	synsets	words	wup_similarity


In [40]:
next(wn.words('fra'))

'déviationniste'

In [42]:
len([x for x in wn.words('jpn')])

89637

In [14]:
from nltk.corpus import wordnet as wn

from lexis import Synsets, KvSynset, KvLemma


s = Synsets()
len(s)

117659

In [2]:
list(s)[:5]

['able.a.01', 'unable.a.01', 'abaxial.a.01', 'adaxial.a.01', 'acroscopic.a.01']

In [3]:
ss = s['sound.n.01']
ss

KvSynset('sound.n.01')

In [4]:
print(*sorted(ss), sep='\n')

also_sees
attributes
causes
definition
entailments
examples
frame_ids
hypernym_distances
hypernym_paths
hypernyms
hyponyms
in_region_domains
in_topic_domains
in_usage_domains
instance_hypernyms
instance_hyponyms
lemma_names
lemmas
lexname
max_depth
member_holonyms
member_meronyms
min_depth
name
offset
part_holonyms
part_meronyms
pos
region_domains
root_hypernyms
similar_tos
substance_holonyms
substance_meronyms
topic_domains
usage_domains
verb_groups


You can retrieve all the key's value at once like this:

In [58]:
dict(ss)

{'attributes': [],
 'usage_domains': [],
 'region_domains': [],
 'definition': 'the particular auditory effect produced by a given cause',
 'entailments': [],
 'member_holonyms': [],
 'hypernym_paths': [[KvSynset('entity.n.01'),
   KvSynset('abstraction.n.06'),
   KvSynset('attribute.n.02'),
   KvSynset('property.n.02'),
   KvSynset('sound_property.n.01'),
   KvSynset('sound.n.01')]],
 'in_usage_domains': [],
 'verb_groups': [],
 'pos': 'n',
 'instance_hypernyms': [],
 'member_meronyms': [],
 'instance_hyponyms': [],
 'hypernym_distances': {(KvSynset('abstraction.n.06'), 4),
  (KvSynset('attribute.n.02'), 3),
  (KvSynset('entity.n.01'), 5),
  (KvSynset('property.n.02'), 2),
  (KvSynset('sound.n.01'), 0),
  (KvSynset('sound_property.n.01'), 1)},
 'topic_domains': [],
 'part_holonyms': [],
 'lexname': 'noun.attribute',
 'root_hypernyms': [Synset('entity.n.01')],
 'causes': [],
 'substance_holonyms': [],
 'examples': ['the sound of rain on the roof', 'the beautiful sound of music'],
 'lem

Let's have a look at a concept with a bit more data

In [56]:
for k, v in sorted(s['dog.n.01'].items()): 
    if v:
        print(f"{k:<16}: {v}\n")

definition      : a member of the genus Canis (probably descended from the common wolf) that has been domesticated by man since prehistoric times; occurs in many breeds

examples        : ['the dog barked all night']

hypernym_distances: {(KvSynset('object.n.01'), 6), (KvSynset('animal.n.01'), 7), (KvSynset('dog.n.01'), 0), (KvSynset('carnivore.n.01'), 2), (KvSynset('entity.n.01'), 13), (KvSynset('living_thing.n.01'), 4), (KvSynset('organism.n.01'), 3), (KvSynset('physical_entity.n.01'), 7), (KvSynset('object.n.01'), 11), (KvSynset('whole.n.02'), 10), (KvSynset('vertebrate.n.01'), 5), (KvSynset('whole.n.02'), 5), (KvSynset('mammal.n.01'), 4), (KvSynset('domestic_animal.n.01'), 1), (KvSynset('animal.n.01'), 2), (KvSynset('living_thing.n.01'), 9), (KvSynset('chordate.n.01'), 6), (KvSynset('organism.n.01'), 8), (KvSynset('physical_entity.n.01'), 12), (KvSynset('placental.n.01'), 3), (KvSynset('entity.n.01'), 8), (KvSynset('canine.n.02'), 1)}

hypernym_paths  : [[KvSynset('entity.n.01'), K

In [5]:
ss = s['dog.n.01']
ss['lemmas']

[Lemma('dog.n.01.dog'),
 Lemma('dog.n.01.domestic_dog'),
 Lemma('dog.n.01.Canis_familiaris')]

In [1]:
lm = KvLemma.from_name('vocal.a.01.vocal')
print(len(dict(lm)))
sorted(lm.dict_of_non_empty_values().items())

33


[('antonyms', [Lemma('instrumental.a.01.instrumental')]),
 ('count', 1),
 ('derivationally_related_forms', [Lemma('vocalize.v.02.vocalize')]),
 ('key', 'vocal%3:01:02::'),
 ('lang', 'eng'),
 ('name', 'vocal'),
 ('pertainyms', [Lemma('voice.n.02.voice')]),
 ('synset', Synset('vocal.a.01')),
 ('syntactic_marker', None)]

In [1]:
from lexis import Synsets, KvSynset, KvLemma

lm = KvSynset.from_name('think.v.01')['lemmas']
lm

[KvLemma('think'),
 KvLemma('believe'),
 KvLemma('consider'),
 KvLemma('conceive')]

In [27]:
lm = KvLemma.from_name('think.v.01.think')
sorted(lm.dict_of_non_empty_values().items())

[('count', 274),
 ('derivationally_related_forms',
  [KvLemma('opinion.n.01.thought'), KvLemma('idea.n.01.thought')]),
 ('frame_ids', [5, 9]),
 ('frame_strings',
  ['Something think something Adjective/Noun', 'Somebody think somebody']),
 ('key', 'think%2:31:01::'),
 ('lang', 'eng'),
 ('name', 'think'),
 ('synset', KvSynset('think.v.01')),
 ('syntactic_marker', None)]

In [8]:
dir(lm)

['_MutableMapping__marker',
 '__abstractmethods__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__slots__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_data_of_obj',
 '_explicit_keys',
 '_from_name',
 '_id_of_key',
 '_ipython_key_completions_',
 '_key_of_id',
 '_keys_cache',
 '_max_repr_size',
 '_obj_of_data',
 '_updatable_cache',
 '_wrapped_methods',
 'clear',
 'dict_of_non_empty_values',
 'from_name',
 'get',
 'head',
 'items',
 'keys',
 'pop',
 'popitem',
 'setdefault',
 'store',
 'update',
 'update_keys_cache',
 'values']

In [25]:
t = lm.derivationally_related_forms()[0]
t

Lemma('opinion.n.01.thought')

In [24]:
for a in dir(t):
    a_val = getattr(t, a)
    if isinstance(a_val, str) and a_val == 'opinion.n.01.thought':
        print(a)

In [23]:
getattr(t, a) == 'opinion.n.01.thought'

AttributeError: 'str' object has no attribute '_name'

In [11]:
lm._synset

Synset('think.v.01')

In [63]:
from nltk.corpus.reader.wordnet import Lemma

dir(Lemma)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__slots__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_frame_ids',
 '_frame_strings',
 '_hypernyms',
 '_instance_hypernyms',
 '_key',
 '_lang',
 '_lex_id',
 '_lexname_index',
 '_name',
 '_related',
 '_synset',
 '_syntactic_marker',
 '_wordnet_corpus_reader',
 'also_sees',
 'antonyms',
 'attributes',
 'causes',
 'count',
 'derivationally_related_forms',
 'entailments',
 'frame_ids',
 'frame_strings',
 'hypernyms',
 'hyponyms',
 'in_region_domains',
 'in_topic_domains',
 'in_usage_domains',
 'instance_hypernyms',
 'instance_hyponyms',
 'key',
 'lang',
 'member_holonyms',
 'member_meronyms',
 'name',
 'part_holonyms',
 'part_meronyms',
 'pertainyms',
 'region_dom

In [70]:
from py2store.sources import Attrs
from i2.signatures import Sig

for k, a in Attrs(lemma).items():
    if callable(a.src):
        print(f"{k}: {Sig(a.src)}")
    else:
        print(f"--> {k}: {a}")

attributes: ()
usage_domains: ()
region_domains: ()
entailments: ()
member_holonyms: ()
in_usage_domains: ()
verb_groups: ()
instance_hypernyms: ()
member_meronyms: ()
topic_domains: ()
key: ()
part_holonyms: ()
count: ()
frame_ids: ()
frame_strings: ()
pertainyms: ()
causes: ()
substance_holonyms: ()
part_meronyms: ()
synset: ()
in_region_domains: ()
derivationally_related_forms: ()
also_sees: ()
hypernyms: ()
name: ()
substance_meronyms: ()
antonyms: ()
hyponyms: ()
lang: ()
similar_tos: ()
syntactic_marker: ()
in_topic_domains: ()
instance_hyponyms: ()


Lemma('dog.n.01.dog')

In [72]:
for k, a in Attrs(lemma).items():
    if callable(a.src):
        print(f"{k}: {a.src()}")

attributes: []
usage_domains: []
region_domains: []
entailments: []
member_holonyms: []
in_usage_domains: []
verb_groups: []
instance_hypernyms: []
member_meronyms: []
topic_domains: []
key: dog%1:05:00::
part_holonyms: []
count: 42
frame_ids: []
frame_strings: []
pertainyms: []
causes: []
substance_holonyms: []
part_meronyms: []
synset: Synset('dog.n.01')
in_region_domains: []
derivationally_related_forms: []
also_sees: []
hypernyms: []
name: dog
substance_meronyms: []
antonyms: []
hyponyms: []
lang: eng
similar_tos: []
syntactic_marker: None
in_topic_domains: []
instance_hyponyms: []


In [56]:
for k, v in sorted(s['dog.n.01'].items()): 
    if v:
        print(f"{k:<16}: {v}\n")

definition      : a member of the genus Canis (probably descended from the common wolf) that has been domesticated by man since prehistoric times; occurs in many breeds

examples        : ['the dog barked all night']

hypernym_distances: {(KvSynset('object.n.01'), 6), (KvSynset('animal.n.01'), 7), (KvSynset('dog.n.01'), 0), (KvSynset('carnivore.n.01'), 2), (KvSynset('entity.n.01'), 13), (KvSynset('living_thing.n.01'), 4), (KvSynset('organism.n.01'), 3), (KvSynset('physical_entity.n.01'), 7), (KvSynset('object.n.01'), 11), (KvSynset('whole.n.02'), 10), (KvSynset('vertebrate.n.01'), 5), (KvSynset('whole.n.02'), 5), (KvSynset('mammal.n.01'), 4), (KvSynset('domestic_animal.n.01'), 1), (KvSynset('animal.n.01'), 2), (KvSynset('living_thing.n.01'), 9), (KvSynset('chordate.n.01'), 6), (KvSynset('organism.n.01'), 8), (KvSynset('physical_entity.n.01'), 12), (KvSynset('placental.n.01'), 3), (KvSynset('entity.n.01'), 8), (KvSynset('canine.n.02'), 1)}

hypernym_paths  : [[KvSynset('entity.n.01'), K

In [56]:
for k, v in sorted(s['dog.n.01'].items()): 
    if v:
        print(f"{k:<16}: {v}\n")

definition      : a member of the genus Canis (probably descended from the common wolf) that has been domesticated by man since prehistoric times; occurs in many breeds

examples        : ['the dog barked all night']

hypernym_distances: {(KvSynset('object.n.01'), 6), (KvSynset('animal.n.01'), 7), (KvSynset('dog.n.01'), 0), (KvSynset('carnivore.n.01'), 2), (KvSynset('entity.n.01'), 13), (KvSynset('living_thing.n.01'), 4), (KvSynset('organism.n.01'), 3), (KvSynset('physical_entity.n.01'), 7), (KvSynset('object.n.01'), 11), (KvSynset('whole.n.02'), 10), (KvSynset('vertebrate.n.01'), 5), (KvSynset('whole.n.02'), 5), (KvSynset('mammal.n.01'), 4), (KvSynset('domestic_animal.n.01'), 1), (KvSynset('animal.n.01'), 2), (KvSynset('living_thing.n.01'), 9), (KvSynset('chordate.n.01'), 6), (KvSynset('organism.n.01'), 8), (KvSynset('physical_entity.n.01'), 12), (KvSynset('placental.n.01'), 3), (KvSynset('entity.n.01'), 8), (KvSynset('canine.n.02'), 1)}

hypernym_paths  : [[KvSynset('entity.n.01'), K

In [88]:
v.tree(lambda v: v.meronyms())

AttributeError: 'Synset' object has no attribute 'meronyms'

In [56]:
for k, v in sorted(s['dog.n.01'].items()): 
    if v:
        print(f"{k:<16}: {v}\n")

definition      : a member of the genus Canis (probably descended from the common wolf) that has been domesticated by man since prehistoric times; occurs in many breeds

examples        : ['the dog barked all night']

hypernym_distances: {(KvSynset('object.n.01'), 6), (KvSynset('animal.n.01'), 7), (KvSynset('dog.n.01'), 0), (KvSynset('carnivore.n.01'), 2), (KvSynset('entity.n.01'), 13), (KvSynset('living_thing.n.01'), 4), (KvSynset('organism.n.01'), 3), (KvSynset('physical_entity.n.01'), 7), (KvSynset('object.n.01'), 11), (KvSynset('whole.n.02'), 10), (KvSynset('vertebrate.n.01'), 5), (KvSynset('whole.n.02'), 5), (KvSynset('mammal.n.01'), 4), (KvSynset('domestic_animal.n.01'), 1), (KvSynset('animal.n.01'), 2), (KvSynset('living_thing.n.01'), 9), (KvSynset('chordate.n.01'), 6), (KvSynset('organism.n.01'), 8), (KvSynset('physical_entity.n.01'), 12), (KvSynset('placental.n.01'), 3), (KvSynset('entity.n.01'), 8), (KvSynset('canine.n.02'), 1)}

hypernym_paths  : [[KvSynset('entity.n.01'), K

In [9]:
dict(s['sound.n.01'])

{'attributes': [],
 'usage_domains': [],
 'region_domains': [],
 'definition': 'the particular auditory effect produced by a given cause',
 'entailments': [],
 'member_holonyms': [],
 'hypernym_paths': [[KvSynset('entity.n.01'),
   KvSynset('abstraction.n.06'),
   KvSynset('attribute.n.02'),
   KvSynset('property.n.02'),
   KvSynset('sound_property.n.01'),
   KvSynset('sound.n.01')]],
 'in_usage_domains': [],
 'verb_groups': [],
 'pos': 'n',
 'instance_hypernyms': [],
 'member_meronyms': [],
 'instance_hyponyms': [],
 'hypernym_distances': {(KvSynset('abstraction.n.06'), 4),
  (KvSynset('attribute.n.02'), 3),
  (KvSynset('entity.n.01'), 5),
  (KvSynset('property.n.02'), 2),
  (KvSynset('sound.n.01'), 0),
  (KvSynset('sound_property.n.01'), 1)},
 'topic_domains': [],
 'part_holonyms': [],
 'lexname': 'noun.attribute',
 'root_hypernyms': [Synset('entity.n.01')],
 'causes': [],
 'substance_holonyms': [],
 'examples': ['the sound of rain on the roof', 'the beautiful sound of music'],
 'lem

['__and__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'isdisjoint']

In [22]:
ss['definition']

'the particular auditory effect produced by a given cause'

In [214]:
Counter(type(getattr(v, x)()) for x in names_of_callable_attributes_that_have_defaults)

Counter({list: 28, str: 4, int: 3, set: 1})

In [211]:
def return_type(x):
    if isinstance(x, list):
        if len(x) == 0:
            return 'empty_list'
        else:
            x0 = x[0]
            if isinstance(x0, list):
                

KvSynset('sound.n.01')

In [212]:
v = s['dog.n.01']
d = defaultdict(list)
for x in names_of_callable_attributes_that_have_defaults:
    a = getattr(v, x)()
    d[type(a)].append((x, a))
d

defaultdict(list,
            {list: [('in_topic_domains', []),
              ('member_holonyms', [Synset('canis.n.01'), Synset('pack.n.06')]),
              ('lemma_names', ['dog', 'domestic_dog', 'Canis_familiaris']),
              ('causes', []),
              ('also_sees', []),
              ('lemmas',
               [Lemma('dog.n.01.dog'),
                Lemma('dog.n.01.domestic_dog'),
                Lemma('dog.n.01.Canis_familiaris')]),
              ('instance_hyponyms', []),
              ('frame_ids', []),
              ('verb_groups', []),
              ('similar_tos', []),
              ('substance_holonyms', []),
              ('root_hypernyms', [Synset('entity.n.01')]),
              ('hyponyms',
               [Synset('basenji.n.01'),
                Synset('corgi.n.01'),
                Synset('cur.n.01'),
                Synset('dalmatian.n.02'),
                Synset('great_pyrenees.n.01'),
                Synset('griffon.n.02'),
                Synset('hunting_dog.

In [183]:
list(s)[1000:1010]

['on_tap.s.02',
 'on_tap.s.01',
 'open.s.06',
 'purchasable.s.02',
 'ready.s.02',
 'unavailable.a.01',
 'inaccessible.s.02',
 'out_of_stock.s.01',
 'awake.a.01',
 'astir.s.01']

In [184]:
'sound.n.01' in s

True

In [185]:
v = s['sound.n.01']

In [186]:
v = s['sound.n.01']
v

KvSynset('sound.n.01')

In [187]:
list(v)

['in_topic_domains',
 'member_holonyms',
 'definition',
 'lemma_names',
 'causes',
 'also_sees',
 'pos',
 'lemmas',
 'instance_hyponyms',
 'frame_ids',
 'verb_groups',
 'similar_tos',
 'substance_holonyms',
 'root_hypernyms',
 'hyponyms',
 'max_depth',
 'part_holonyms',
 'topic_domains',
 'offset',
 'examples',
 'part_meronyms',
 'in_usage_domains',
 'instance_hypernyms',
 'lexname',
 'min_depth',
 'hypernym_distances',
 'region_domains',
 'member_meronyms',
 'in_region_domains',
 'usage_domains',
 'hypernyms',
 'entailments',
 'hypernym_paths',
 'name',
 'attributes',
 'substance_meronyms']

In [198]:
dict(v)

{'in_topic_domains': [],
 'member_holonyms': [],
 'definition': 'the particular auditory effect produced by a given cause',
 'lemma_names': ['sound'],
 'causes': [],
 'also_sees': [],
 'pos': 'n',
 'lemmas': [Lemma('sound.n.01.sound')],
 'instance_hyponyms': [],
 'frame_ids': [],
 'verb_groups': [],
 'similar_tos': [],
 'substance_holonyms': [],
 'root_hypernyms': [Synset('entity.n.01')],
 'hyponyms': [KvSynset('noisiness.n.01'),
  KvSynset('ring.n.01'),
  KvSynset('unison.n.03'),
  KvSynset('voice.n.01')],
 'max_depth': 5,
 'part_holonyms': [],
 'topic_domains': [],
 'offset': 4981139,
 'examples': ['the sound of rain on the roof', 'the beautiful sound of music'],
 'part_meronyms': [],
 'in_usage_domains': [],
 'instance_hypernyms': [],
 'lexname': 'noun.attribute',
 'min_depth': 5,
 'hypernym_distances': {(Synset('abstraction.n.06'), 4),
  (Synset('attribute.n.02'), 3),
  (Synset('entity.n.01'), 5),
  (Synset('property.n.02'), 2),
  (KvSynset('sound.n.01'), 0),
  (Synset('sound_prope

In [199]:
v['hypernym_paths']

[[Synset('entity.n.01'),
  Synset('abstraction.n.06'),
  Synset('attribute.n.02'),
  Synset('property.n.02'),
  Synset('sound_property.n.01'),
  KvSynset('sound.n.01')]]

In [191]:
v.hypernym_paths()

[[Synset('entity.n.01'),
  Synset('abstraction.n.06'),
  Synset('attribute.n.02'),
  Synset('property.n.02'),
  Synset('sound_property.n.01'),
  KvSynset('sound.n.01')]]

In [195]:
dict(v)

in_topic_domains
member_holonyms
definition
lemma_names
causes
also_sees
pos
lemmas
instance_hyponyms
frame_ids
verb_groups
similar_tos
substance_holonyms
root_hypernyms
hyponyms
max_depth
part_holonyms
topic_domains
offset
examples
part_meronyms
in_usage_domains
instance_hypernyms
lexname
min_depth
hypernym_distances
region_domains
member_meronyms
in_region_domains
usage_domains
hypernyms
entailments
hypernym_paths
name
attributes
substance_meronyms


{'in_topic_domains': [],
 'member_holonyms': [],
 'definition': 'the particular auditory effect produced by a given cause',
 'lemma_names': ['sound'],
 'causes': [],
 'also_sees': [],
 'pos': 'n',
 'lemmas': [Lemma('sound.n.01.sound')],
 'instance_hyponyms': [],
 'frame_ids': [],
 'verb_groups': [],
 'similar_tos': [],
 'substance_holonyms': [],
 'root_hypernyms': [Synset('entity.n.01')],
 'hyponyms': [KvSynset('noisiness.n.01'),
  KvSynset('ring.n.01'),
  KvSynset('unison.n.03'),
  KvSynset('voice.n.01')],
 'max_depth': 5,
 'part_holonyms': [],
 'topic_domains': [],
 'offset': 4981139,
 'examples': ['the sound of rain on the roof', 'the beautiful sound of music'],
 'part_meronyms': [],
 'in_usage_domains': [],
 'instance_hypernyms': [],
 'lexname': 'noun.attribute',
 'min_depth': 5,
 'hypernym_distances': {(Synset('abstraction.n.06'), 4),
  (Synset('attribute.n.02'), 3),
  (Synset('entity.n.01'), 5),
  (Synset('property.n.02'), 2),
  (KvSynset('sound.n.01'), 0),
  (Synset('sound_prope

In [30]:
v.lemma_names(lang='fra')

['son']

In [33]:
v.lemma_names(lang='jpn')

['サウンド', '音', '音声', '音響', '響', '響き']

In [40]:
from lexis import Synsets

Synsets

lexis.Synsets

In [39]:
[a for a in Attrs(v) if callable(getattr(v, a)) and len(Sig(getattr(v, a))) == 0]

NameError: name 'Attrs' is not defined

In [124]:
Synset

[autoreload of lexis failed: Traceback (most recent call last):
  File "/Users/twhalen/.virtualenvs/382/lib/python3.8/site-packages/IPython/extensions/autoreload.py", line 245, in check
    superreload(m, reload, self.old_objects)
  File "/Users/twhalen/.virtualenvs/382/lib/python3.8/site-packages/IPython/extensions/autoreload.py", line 394, in superreload
    module = reload(module)
  File "/Users/twhalen/.pyenv/versions/3.8.2/lib/python3.8/imp.py", line 314, in reload
    return importlib.reload(module)
  File "/Users/twhalen/.pyenv/versions/3.8.2/lib/python3.8/importlib/__init__.py", line 169, in reload
    _bootstrap._exec(spec, module)
  File "<frozen importlib._bootstrap>", line 604, in _exec
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/D/Dropbox/dev/p3/proj/i/py2store/lexis.py", line 35, in <module>
    from lexis import Synsets
ImportError: cannot import name 'Sy

nltk.corpus.reader.wordnet.Synset

In [43]:
from py2store.sources import Attrs
from lexis import Synsets

w = Attrs(v)
list(w)

NameError: name 'v' is not defined

In [125]:
import lexis

lexis


In [ ]:
[a for a in Attrs(v) if callable(getattr(v, a)) and len(Sig(getattr(v, a))) == 0]


In [67]:
from i2.signatures import Sig
from py2store.sources import Attrs
from nltk.corpus.reader.wordnet import Synset

[name for name, a in Attrs(Synset).items() if callable(a.src) and len(Sig(a.src)) == 1]

['in_topic_domains',
 'member_holonyms',
 'definition',
 'causes',
 'also_sees',
 'pos',
 'instance_hyponyms',
 'frame_ids',
 'verb_groups',
 'similar_tos',
 'root_hypernyms',
 'substance_holonyms',
 'hyponyms',
 'max_depth',
 'part_holonyms',
 'topic_domains',
 'offset',
 'examples',
 'part_meronyms',
 'in_usage_domains',
 'instance_hypernyms',
 'lexname',
 'min_depth',
 'region_domains',
 'member_meronyms',
 'in_region_domains',
 'usage_domains',
 'hypernyms',
 'entailments',
 'hypernym_paths',
 'name',
 'attributes',
 'substance_meronyms']

In [69]:
for name, a in Attrs(Synset).items():
    if callable(a.src) and len(Sig(a.src)) > 1:
        print(f"{name}{Sig(a.src)}")

res_similarity(self, other, ic, verbose=False)
lemma_names(self, lang='eng')
lemmas(self, lang='eng')
lin_similarity(self, other, ic, verbose=False)
closure(self, rel, depth=-1)
lch_similarity(self, other, verbose=False, simulate_root=True)
lowest_common_hypernyms(self, other, simulate_root=False, use_min_depth=False)
common_hypernyms(self, other)
wup_similarity(self, other, verbose=False, simulate_root=True)
hypernym_distances(self, distance=0, simulate_root=False)
shortest_path_distance(self, other, simulate_root=False)
tree(self, rel, depth=-1, cut_mark=None)
path_similarity(self, other, verbose=False, simulate_root=True)
jcn_similarity(self, other, ic, verbose=False)


In [107]:
s['sound.n.01']


Synset('sound.n.01')

In [110]:
Sig.defaults
t = next(iter(Attrs(Synset).values()))
Sig(t)

<Sig ()>

In [123]:
names_of_non_callable_attributes = {name for name, a in Attrs(Synset).items() if not callable(a.src)}

names_of_callable_attributes_that_have_defaults = {
    name for name, a in Attrs(Synset).items() 
    if callable(a.src) and len(Sig(a.src)) - len(Sig(a.src).defaults) == 1}

kv_synset_key_names = names_of_non_callable_attributes | names_of_callable_attributes_that_have_defaults

from dol import Store, KvReader

class KvSynset(Store, Synset):
    key_names = kv_synset_key_names
    
    def __getitem__(self, k):
        attr = getattr(self, k)
        if callable(attr):
            return attr()
        else:
            return attr
        
    def __iter__(self):
        yield from key_names
        
    def __contains__(self, k):
        return k in self.key_names
    
    def __len__(self, k):
        return len()
    
    def __repr__(self):
        return f"{self.__class__.__name__}('{self._name}')"

    
KvSynset(v)

KvSynset('sound.n.01')

In [113]:
sig = Sig(v.shortest_path_distance)
len(sig) - len(sig.defaults)

1

In [98]:
from py2store import Store

class KvSynset(Synset, Store):
    pass

Synset('sound.n.01')

Synset('None')

In [102]:
Synset.

Synset('sound.n.01')

In [72]:
s['sound.n.01'].lch_similarity(s['sound.n.02'])

1.1526795099383855

In [78]:
s['sound.n.01'].hypernym_distances()

{(Synset('abstraction.n.06'), 4),
 (Synset('attribute.n.02'), 3),
 (Synset('entity.n.01'), 5),
 (Synset('property.n.02'), 2),
 (Synset('sound.n.01'), 0),
 (Synset('sound_property.n.01'), 1)}

In [85]:
v.hypernyms()

[Synset('sound_property.n.01')]

In [ ]:
v.

In [94]:
v.tree(lambda v: v.hyponyms(), depth=2)

[Synset('sound.n.01'),
 [Synset('noisiness.n.01'), [Synset('boisterousness.n.02')]],
 [Synset('ring.n.01')],
 [Synset('unison.n.03')],
 [Synset('voice.n.01'), [Synset('androglossia.n.01')]]]

In [83]:
s['sound.n.01'].common_hypernyms(s['sound.n.02'])

[Synset('entity.n.01'), Synset('abstraction.n.06')]

In [ ]:
s['sound.n.01'].lch_similarity(s['sound.n.02'])

In [54]:
list(Attrs(Synset))

['in_topic_domains',
 'member_holonyms',
 'res_similarity',
 'definition',
 'lemma_names',
 'causes',
 'also_sees',
 'pos',
 'lemmas',
 'instance_hyponyms',
 'lin_similarity',
 'frame_ids',
 'verb_groups',
 'closure',
 'similar_tos',
 'lch_similarity',
 'root_hypernyms',
 'lowest_common_hypernyms',
 'substance_holonyms',
 'hyponyms',
 'max_depth',
 'common_hypernyms',
 'part_holonyms',
 'topic_domains',
 'offset',
 'examples',
 'wup_similarity',
 'part_meronyms',
 'in_usage_domains',
 'instance_hypernyms',
 'lexname',
 'min_depth',
 'hypernym_distances',
 'region_domains',
 'member_meronyms',
 'in_region_domains',
 'shortest_path_distance',
 'tree',
 'path_similarity',
 'usage_domains',
 'hypernyms',
 'entailments',
 'hypernym_paths',
 'name',
 'attributes',
 'substance_meronyms',
 'jcn_similarity']

In [58]:
t = Attrs(Synset)['attributes']
t.src

<function nltk.corpus.reader.wordnet._WordNetObject.attributes(self)>

In [ ]:
wn.words()


In [17]:
from i2.signatures import Sig
from py2store.sources import Attrs
[a for a in Attrs(v) if len(Sig(getattr(v, a))) == 0]


['min_depth',
 'region_domains',
 'topic_domains',
 'root_hypernyms',
 'hyponyms',
 'causes',
 'pos',
 'part_holonyms',
 'max_depth',
 'entailments',
 'hypernym_paths',
 'in_usage_domains',
 'substance_holonyms',
 'member_meronyms',
 'instance_hyponyms',
 'similar_tos',
 'examples',
 'frame_ids',
 'in_topic_domains',
 'also_sees',
 'hypernyms',
 'instance_hypernyms',
 'member_holonyms',
 'substance_meronyms',
 'name',
 'part_meronyms',
 'usage_domains',
 'offset',
 'definition',
 'verb_groups',
 'in_region_domains',
 'attributes',
 'lexname']

In [8]:
print(f"{v.definition()=}")
print(f"{v.hyponyms()=}")
print(f"{v.hypernyms()=}")
print(f"{v.also_sees()=}")
print(f"{v.also_sees()=}")

v.definition()='the particular auditory effect produced by a given cause'
v.hyponyms()=[Synset('noisiness.n.01'), Synset('ring.n.01'), Synset('unison.n.03'), Synset('voice.n.01')]
v.hypernyms()=[Synset('sound_property.n.01')]
v.also_sees()=[]
